Lib/test/test_logging.py

```diff
@@ -4310,6 +4310,24 @@ def test_issue_89047(self):
             s = f.format(r)
             self.assertNotIn('.1000', s)
 
+    def test_issue_102402_100msecs(self):
+        # quick hack for sanity check
+        og_time_ns = time.time_ns
+        # (time_ns, expected msecs value)
+        tests = (
+            (1_677_902_297_100_000_000, 100.0),  # exactly 100ms
+            (1_677_903_920_999_998_503, 999.0),  # rounding up
+            (1_677_903_920_000_998_503, 0.0),  # rounding up
+
+        )
+        try:
+            for ns, want in tests:
+                time.time_ns = lambda: ns
+                record = logging.makeLogRecord({'msg': 'test'})
+                self.assertEqual(record.msecs, want)
+        finally:
+            time.time_ns = og_time_ns
```

Comment:
You can use `unittest.mock.patch` as in this example:
```python
import time
from unittest.mock import patch

for t in (123, 456, 789):
    with patch('time.time_ns') as patched_ns:
        patched_ns.return_value = t
        value = time.time_ns()
        print(f'patched: {value}')
    print(f'OG: {time.time_ns()}')
```
which should, when run, print something like this:
```
patched: 123
OG: 1681278183002693921
patched: 456
OG: 1681278183003199141
patched: 789
OG: 1681278183003610801
```

Reply:
Ah, my assumption was that we probably didn't want to add imports (I don't know why I assumed that...).

Updated to patch with `mock` instead of manual patching.